### PR TITLE
Convert CallbackQueue to a class

### PR DESCRIPTION
**Pull request by vjeux**
It turns out that flow cannot type `this` with a function constructor + prototype overrides. Turning it to a class makes flow happy and has minimal impact on the output.

In open source, we already use the loose version of the class transform and internally we have one that's outputting even less code if you have `@preventMunge` in the header.

See discussion in https://www.facebook.com/groups/2003630259862046/permalink/2098480820376989/

### CODE REVIEW

**Comment by vjeux** [comment]
Flow coverage goes from 44% where all the `this` are untyped (and therefore all the methods called on it and their return values) to 96%. The only missing piece is `invariant` which is not covered for some reason

**Comment by vjeux** [comment]
cc @spicyj

File: src/renderers/shared/utils/CallbackQueue.js [hunk 1]
```diff
@@ -41,23 +43,23 @@ Object.assign(CallbackQueue.prototype, {
    * @param {?object} context Context to call `callback` with.
    * @internal
    */
-  enqueue: function(callback, context) {
+   enqueue(callback: () => void, context: T) {
     this._callbacks = this._callbacks || [];
-    this._contexts = this._contexts || [];
     this._callbacks.push(callback);
+    this._contexts = this._contexts || [];
```
Comment (spicyj) [inline src/renderers/shared/utils/CallbackQueue.js]: Why did you move this?

Comment (spicyj) [inline src/renderers/shared/utils/CallbackQueue.js]: Never mind, I guess I see.

**Comment by spicyj** [comment]
lgtm

File: src/renderers/shared/utils/CallbackQueue.js (hunk 1 above)
Comment (vjeux) [inline src/renderers/shared/utils/CallbackQueue.js]: Yeah, flow doesn't know that push() isn't messing up with everything. I figured moving it one line after isn't a big deal to make flow happy

**Comment by vjeux** [comment]
Grrr I'm bad at running eslint, I'll send an update fixing lint

**Comment by vjeux** [comment]
Fix lint errors

File: src/renderers/shared/stack/event/ReactSyntheticEvent.js [hunk 2]
```diff
@@ -18,4 +18,4 @@ export type DispatchConfig = any;
 
 export class ReactSyntheticEvent extends SyntheticEvent {
   dispatchConfig: DispatchConfig;
-};
+}
```
Comment (flarnie) [inline src/renderers/shared/stack/event/ReactSyntheticEvent.js]: Thanks for catching that!
